### PR TITLE
admin: add drain listeners endpoint 

### LINKED ISSUE #4460: Add ability to drain the listeners using the admin interface
*Title*: Add ability to drain the listeners using the admin interface

*Description*:
Drain the listeners on demand, with the assumption that we're going to shutdown shortly afterwards, so future updates to those listeners should be ignored.

cc @nmittler

### PR DESCRIPTION
**Pull request by ramaraochavali**
Description: Adds an admin endpoint that allows listeners to be drained.
Risk Level: Low, only admin initiated.
Testing: Added tests
Docs Changes: Updated
Release Notes: Added
Fixes #4460

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@mattklein123 PTAL

File: test/integration/integration_admin_test.cc [hunk 1]
```diff
@@ -122,6 +122,12 @@ std::string ContentType(const BufferingStreamDecoderPtr& response) {
 } // namespace
 
 TEST_P(IntegrationAdminTest, Admin) {
+
```
Comment (mattklein123) [inline test/integration/integration_admin_test.cc]: nit: remove extra newline here and below.

File: test/integration/integration_admin_test.cc [hunk 2]
```diff
@@ -414,6 +420,42 @@ TEST_P(IntegrationAdminTest, Admin) {
...
+  EXPECT_EQ("text/plain; charset=UTF-8", ContentType(response));
+  EXPECT_EQ(0, test_server_->counter("listener_manager.listener_removed")->value());
+
+  // Validate that the outbound listener is drained on drain listeners.
+  response = IntegrationUtil::makeSingleRequest(lookupPort("admin"), "POST", "/drain_listeners", "",
+                                                downstreamProtocol(), version_);
+  EXPECT_TRUE(response->complete());
+  EXPECT_EQ("200", response->headers().Status()->value().getStringView());
+  EXPECT_EQ("text/plain; charset=UTF-8", ContentType(response));
+  test_server_->waitForCounterGe("listener_manager.listener_removed", 1);
```
Comment (mattklein123) [inline test/integration/integration_admin_test.cc]: How long do these tests take to run with real time? I'm guessing many seconds because of draining? I think you will likely need to modify the drain time for the listener or Envoy to make this shorter?

More generally, can you potentially do a test with actual requests to make sure that the listener is actually draining and then gone when we do this admin command?

Comment (ramaraochavali) [reply test/integration/integration_admin_test.cc]: It runs for 1s.  For integration test server it is set here https://github.com/envoyproxy/envoy/blob/master/test/integration/server.cc#L42.

My intention was not not test the actual draining behaviour - but ensure that it is triggered as draining api would have already been tested. Added a new test now. Please see if that is what you are expecting?

File: source/server/http/admin.cc [hunk 3]
```diff
@@ -686,6 +686,20 @@ Http::Code AdminImpl::handlerMemory(absl::string_view, Http::HeaderMap& response
   return Http::Code::OK;
 }
 
+Http::Code AdminImpl::handlerDrainListeners(absl::string_view url, Http::HeaderMap&,
+                                            Buffer::Instance& response, AdminStream&) {
+  const Http::Utility::QueryParams params = Http::Utility::parseQueryString(url);
+  const bool inbound_only = params.find("inboundonly") != params.end();
+  for (const Network::ListenerConfig& listener : server_.listenerManager().listeners()) {
+    if (!inbound_only || listener.direction() == envoy::api::v2::core::TrafficDirection::INBOUND) {
+      server_.listenerManager().removeListener(listener.name());
```
Comment (mattklein123) [inline source/server/http/admin.cc]: In thinking about this more, I'm not convinced this is correct. I think if you remove the listener here, it's possible that LDS will add it back? Don't we want to stop the listener instead? WDYT?

Comment (ramaraochavali) [reply source/server/http/admin.cc]: I think assuming that it is operationally handled that this is invoked on SIGTERM and wait for drain time is done and then Envoy goes away any ways, it may not be a big issue because till the listener is draining it will not allow to add back?

Are you suggesting, we call the drain completion with empty call back here https://github.com/envoyproxy/envoy/blob/master/source/server/listener_manager_impl.cc#L668 for this case so that it won't remove?

Let me know if you want to handle that case - we will have to introduce a new API to listener manager called stopListener and call drain with null calback? or any other better way?

Comment (ramaraochavali) [reply source/server/http/admin.cc]: Went ahead and added api for `stopListener`. Please see if this better.

File: source/server/listener_manager_impl.cc [hunk 4]
```diff
@@ -767,6 +770,38 @@ uint64_t ListenerManagerImpl::numConnections() {
   return num_connections;
 }
 
+bool ListenerManagerImpl::stopListener(const std::string& name) {
```
Comment (mattklein123) [inline source/server/listener_manager_impl.cc]: I'm wondering if we can simplify this change by just making this API take a `ListenerConfig`, and then just call the worker stop listener API directly with that config? I think this will work correctly.

I think no matter we have a potential race condition here in which if we remove anything, LDS will come and add it back later I think. Given this, it seems better to just stop the listeners directly? This would not cover the warming case, though that might be good enough? I think to completely remove any race conditions you would need to kill warming listeners and keep them from coming back by name? I'm really not sure if this type of complexity is worth it, but might be worth a bunch of comments.

WDYT?

**Comment by ramaraochavali** [comment]
I was thinking that as another option. I definitely agree that the
complexity is not worth in this case. I will switch to calling
stopListeners directly

File: source/server/listener_manager_impl.cc [hunk 5]
```diff
@@ -668,12 +668,12 @@ void ListenerManagerImpl::drainListener(ListenerImplPtr&& listener) {
...
     for (const auto& worker : workers_) {
-      // Once the drain time has completed via the drain manager's timer, we tell the workers to
-      // remove the listener.
+      // Once the drain time has completed via the drain manager's timer, we tell the workers
+      // to remove the listener.
       worker->removeListener(*draining_it->listener_, [this, draining_it]() -> void {
-        // The remove listener completion is called on the worker thread. We post back to the main
-        // thread to avoid locking. This makes sure that we don't destroy the listener while filters
-        // might still be using its context (stats, etc.).
+        // The remove listener completion is called on the worker thread. We post back to the
```
Comment (ramaraochavali) [inline source/server/listener_manager_impl.cc]: This is not intended, but when I ran `./ci/run_envoy_docker.sh './ci/do_ci.sh fix_format'` - it automatically formatted this.

**Comment by ramaraochavali** [comment]
@mattklein123 Sorry for the delay. I have change the API to stop workers as discussed with a bunch of comments on API. PTAL.

File: include/envoy/server/listener_manager.h [hunk 6]
```diff
@@ -153,6 +153,19 @@ class ListenerManager {
    */
   virtual bool removeListener(const std::string& name) PURE;
 
+  /*
+   * Stops a listener from accepting new connections without actually removing any of them. This is
```
Comment (mattklein123) [inline include/envoy/server/listener_manager.h]: Thanks @ramaraochavali for iterating on this. IMO it's OK to ship with the warming / config change race condition as it's well documented. To that effect, can you actually move some of this documentation into the admin handler docs above under an `.. attention::` block?

Also, @alyssawilk @lizan @PiotrSikora any concerns with shipping with this known race condition until if/when someone complains about it? (Assuming it is well documented.)

/wait

Comment (ramaraochavali) [reply include/envoy/server/listener_manager.h]: move the docs to admin handler docs. PTAL

Comment (PiotrSikora) [reply include/envoy/server/listener_manager.h]: Since the goal of this feature is to help with graceful shutdown, the "stop/drain listeners" operation should be a permanent condition, and allowing LDS to add new listeners would render it a bit useless, IMHO.

I think that making this a permanent condition (vs racy) would require adding a boolean to listener manager, which would result in all new LDS resources in a given direction being rejected, and possibly extra check to drain new listeners, immediately after they're warmed, in case they were warming at the time of the request (although, looking at `ListenerManagerImpl::removeListener`, it seems that it should be possible to destroy warming up listeners before they warm up).

File: source/server/http/admin.cc [hunk 7]
```diff
@@ -686,6 +686,26 @@ Http::Code AdminImpl::handlerMemory(absl::string_view, Http::HeaderMap& response
   return Http::Code::OK;
 }
 
+Http::Code AdminImpl::handlerDrainListeners(absl::string_view url, Http::HeaderMap&,
+                                            Buffer::Instance& response, AdminStream&) {
+  const Http::Utility::QueryParams params = Http::Utility::parseQueryString(url);
+  const bool inbound_only = params.find("inboundonly") != params.end();
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply source/server/http/admin.cc]: If you want both it is possible with `/drain_listeners` with out any request params. `inboundonly` would be used during rolling upgrades to stop incoming connections. `outboundonly` can be added later?

File: include/envoy/server/listener_manager.h (hunk 6 above)
Comment (ramaraochavali) [reply include/envoy/server/listener_manager.h]: Another option I can think of if we want to stop LDS to add/remove  is to maintain a list of stopped listeners (not just direction because we may want to stop listener by name in future) and block add/remove and handle in warm up callback if the listener is stopped. I have been experimenting with it. Let me push that change and see if that is better.

Comment (ramaraochavali) [reply include/envoy/server/listener_manager.h]: @mattklein123  pushed that change. Please see if this is better or any other condition is missed?

File: source/server/listener_manager_impl.cc [hunk 8]
```diff
@@ -770,11 +784,19 @@ uint64_t ListenerManagerImpl::numConnections() {
 void ListenerManagerImpl::stopListener(Network::ListenerConfig& listener) {
   ENVOY_LOG(debug, "begin stop listener: name={}", listener.name());
   for (const auto& worker : workers_) {
-    worker->stopListener(listener);
+    if (!listenerStopped(listener.name())) {
+      stopped_listeners_.push_back(listener.name());
```
Comment (PiotrSikora) [inline source/server/listener_manager_impl.cc]: This is nice future-proofing, but it doesn't work for the basic use case of "stop/drain all inbound listeners" (existing and future), we need a direction-based guard for that.

Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: the stopped_listeners_ will have all the inbound listeners of we call drain_listeners?inboundonly - would not that be sufficient?

Comment (PiotrSikora) [reply source/server/listener_manager_impl.cc]: It will only have the inbound listeners that were active at the time of the operation, but LDS can add new inbound listeners afterwards, and if I'm reading this correctly, those listeners will be happily accepted by Envoy.

Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: I see.

Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: My mental model has been, no new ingress listeners will come up after pod/service has come up - they might change but not new ones.

**Comment by ramaraochavali** [comment]
@mattklein123 @PiotrSikora Thinking about this bit more, I think the graceful shutdown case of inbound listeners would still be useful even with race condition because, ingress listeners changing is a rare condition

- New ingress listeners would not come on existing pods, new listeners means service is listening on new ports which is a different version
- Existing ingress listeners change very rarely

Given this, IMO we should ship and iterate here - we can try this and see how it is helping (things would definitely be not worse) and based on operational experience we can tweak in follow-up PRs.

So I have reverted my previous commit. 

WDYT?

**Comment by mattklein123** [comment]
> Given this, IMO we should ship and iterate here - we can try this and see how it is helping (things would definitely be not worse) and based on operational experience we can tweak in follow-up PRs.

I did not fully track the back and forth between you and @PiotrSikora. Can you summarize? Personally, I'm OK shipping with the race condition as long as it is clearly documented. (My assumption is @PiotrSikora commented on the same race condition I did and then asked you to fix it, I'm not sure what happened from there as I see any changes have now been reverted.) With that said, if @PiotrSikora feels strongly that we should not ship something with a race condition, I'm sure we can iterate to a solution. 

/wait-any

**Comment by ramaraochavali** [comment]
@mattklein123 Sorry. may be reverted too early., The summary of the discussion is when draining is initiated from admin endpoint, we should hold that info in listener_manager (the traffic direction on which the listener initiated) and then block add/update/warm/remove for any listener changes in that traffic direction. 
We can certainly try this, but then the drain_listeners would not be very generic in the sense that we can not track draining listener by name for example.
I was really not sure if it is worth without actually trying this - That is why I reverted back to old state. 
Yeah, agree if @PiotrSikora  feels strongly about this, we can iterate to a solution.

**Comment by mattklein123** [comment]
> We can certainly try this, but then the drain_listeners would not be very generic in the sense that we can not track draining listener by name for example.

We could definitely implement blocking by name if we want. This is the rough solution I was thinking of to fix the race condition. In any case, I think we need to step back and decide if we need to handle the race condition in the MVP of this feature or not. @PiotrSikora WDYT?

**Comment by ramaraochavali** [comment]
> We could definitely implement blocking by name if we want.

That is what I implemented by holding the stopped listener names in a list. But then the problem that @PiotrSikora brought up was if a new ingress listener is added - it wont be in the list because by the time `drain_listeners?inboundonly` was invoked, this listener did not exist. So he was thinking blocking by traffic direction. Then at that point, I stepped back and thought should we really do it as MVP for this.

**Comment by lambdai** [comment]
What about adding a `is_stopping` flag into ListenerManagerImpl. ListenerManagerImpl should not warm up listener if `is_stopping`.
It is probably hard to support transit from `is_stopping = false` to `is_stopping = true`, but probably we have to support

**Comment by ramaraochavali** [comment]
> ListenerManagerImpl should not warm up listener if is_stopping.

We can not generically stopping warming all listeners - right?

**Comment by ramaraochavali** [comment]
I will be OOO next 3 days, but will address any feedback or implement any agreed up on solution to eliminate the race if we decide to do when I am back.

**Comment by PiotrSikora** [comment]
@ramaraochavali the difference between what you have with listeners names and what I want should be as simple as adding a check to `listenerStopped` to see if the direction of the new listener is already drained/stopped.

I feel rather strongly about making this a permanent state, since making this a temporary state, and allowing new listeners to be added, will result in hard to diagnose errors in production.

**Comment by mattklein123** [comment]
> I feel rather strongly about making this a permanent state, since making this a temporary state, and allowing new listeners to be added, will result in hard to diagnose errors in production.

Yup this is fine with me. The approach that @PiotrSikora is suggesting is the one I would have suggested also. Let's go with that when you get back. Thank you Rama!

/wait

**Comment by ramaraochavali** [comment]
@mattklein123 @PiotrSikora Changed as discussed. PTAL.

File: source/server/listener_manager_impl.cc [hunk 9]
```diff
@@ -780,6 +788,25 @@ uint64_t ListenerManagerImpl::numConnections() {
...
 
+void ListenerManagerImpl::stopListener(Network::ListenerConfig& listener) {
+  ENVOY_LOG(debug, "begin stop listener: name={}", listener.name());
+  for (const auto& worker : workers_) {
+    if (listener.direction() == envoy::api::v2::core::TrafficDirection::INBOUND) {
+      inbound_listeners_stopped_ = true;
+    }
+    if (listener.direction() == envoy::api::v2::core::TrafficDirection::OUTBOUND) {
+      outbound_listeners_stopped_ = true;
+    }
```
Comment (PiotrSikora) [inline source/server/listener_manager_impl.cc]: I don't think this works. Notably, if you hit `/drain_listeners` to stop all listeners, and all of them have unspecified traffic direction, then neither `inbound_listeners_stopped_` nor `outbound_listeners_stopped_` will be set to true.

File: source/server/listener_manager_impl.h [hunk 10]
```diff
@@ -166,6 +167,15 @@ class ListenerManagerImpl : public ListenerManager, Logger::Loggable<Logger::Id:
...
     stats_.total_listeners_active_.set(active_listeners_.size());
   }
+  bool listnerStopped(const envoy::api::v2::Listener& config) {
+    // Currently all listeners in a given direction are stopped because of the way admin
+    // drain_listener functionality is implemented. So it is OK to check traffic direction. This
+    // needs to be revisited, if that changes - if we support drain by listener name,for example.
+    return (config.traffic_direction() == envoy::api::v2::core::TrafficDirection::INBOUND &&
+            inbound_listeners_stopped_) ||
+           (config.traffic_direction() == envoy::api::v2::core::TrafficDirection::OUTBOUND &&
+            outbound_listeners_stopped_);
```
Comment (PiotrSikora) [inline source/server/listener_manager_impl.h]: Similarly, this requires traffic direction to be configured for `/drain_listeners` to work.

File: source/server/listener_manager_impl.cc (hunk 9 above)
Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: The approach as discussed https://github.com/envoyproxy/envoy/issues/4460#issuecomment-535449831 is based on traffic direction being set to distinguish between inbound vs outbound listeners as that is a simple way to distinguish. Otherwise I am not sure of any other way to easily distinguish. Since it is needed for inbound, we might as well use it for` /drain_listeners` - to stop all listeners instead of using a different approach there. IMO, that is a fair assumption and an easy change on control plane side to set the listener direction -Istio already does it for example.

Comment (PiotrSikora) [reply source/server/listener_manager_impl.cc]: The approach is fine (using configured traffic direction when draining inbound/outbound listeners).

However, the implementation isn't, since it's setting the state based on the configured traffic direction of the listener that's being drained, and not on what was requested, e.g. a request to drain all listeners (using `/drain_listeners`), when none of the listeners have configured traffic direction, won't modify `inbound_listeners_stopped_` and `outbound_listeners_stopped_`, and `listnerStopped()` will always return `false`.

Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: I see. I will make change to ignore stop listener/ not stop listner if traffic_direction is not set (UNSPECIFIED). Is that fine?

Comment (PiotrSikora) [reply source/server/listener_manager_impl.cc]: No, when draining ALL listeners (`/drain_listeners`), traffic direction shouldn't matter. All listeners should be drained, and no new listeners should be accepted using LDS.

Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: OK. Changed it like that. PTAL. I changed the new method to `shutdownListeners` to better describe what we are doing here.

File: source/server/http/admin.cc [hunk 11]
```diff
@@ -697,6 +697,18 @@ Http::Code AdminImpl::handlerMemory(absl::string_view, Http::HeaderMap& response
...
+  if (server_.listenerManager().shutdownListeners(inbound_only)) {
+    response.add("OK\n");
+  } else {
+    response.add("No listeners matched\n");
```
Comment (PiotrSikora) [inline source/server/http/admin.cc]: Nit: I'm not sure if it really matters whether we matched any listeners or not. I'm fine leaving it as-is, but pointing this out in case anyone else cares.

File: include/envoy/server/listener_manager.h [hunk 12]
```diff
@@ -153,6 +153,15 @@ class ListenerManager {
...
   virtual bool removeListener(const std::string& name) PURE;
 
+  /*
+   * Stops listeners from accepting new connections without actually removing them. This is
+   * used by /drain_listeners admin endpoint. This method directly stops the listeners on workers.
+   * Once a listener is stopped, any listener modifications are not allowed.
+   * @param inbound_only if true stops all the inbound listeners and stops all listeners if false.
+   * @return true if any listeners have been stopped.
+   */
+  virtual bool shutdownListeners(bool inbound_only) PURE;
```
Comment (mattklein123) [inline include/envoy/server/listener_manager.h]: Can this API just be merged into `stopListeners()` below? Can that API take an enum of what type of stop to perform?

File: source/server/listener_manager_impl.h [hunk 13]
```diff
@@ -195,6 +204,8 @@ class ListenerManagerImpl : public ListenerManager, Logger::Loggable<Logger::Id:
   std::list<DrainingListener> draining_listeners_;
   std::list<WorkerPtr> workers_;
   bool workers_started_{};
+  bool inbound_listeners_stopped_{};
```
Comment (mattklein123) [inline source/server/listener_manager_impl.h]: switch these 2 bools to a "stop state" enum?

File: include/envoy/server/listener_manager.h (hunk 12 above)
Comment (ramaraochavali) [reply include/envoy/server/listener_manager.h]: We could do that and actually did think about it. But couple of points 
1. Do we have to maintain this api for any backward compatibility reasons because it is a public API?
2. When you mean stop state enum - are you thinking AllListeners, Inbound Listeners as values? I was thinking we could introduce a more generic “StopListenerSelector” that can have more wide categories of criteria like name etc instead of an enum - And if we want to do that, can we do that later? 
If you think more wide criteria is not needed at all, I can go with enum and change it now. Let me know

File: source/server/listener_manager_impl.h (hunk 13 above)
Comment (ramaraochavali) [reply source/server/listener_manager_impl.h]: Related to my other comment above. I can change based on what we decide there

File: include/envoy/server/listener_manager.h (hunk 12 above)
Comment (ramaraochavali) [reply include/envoy/server/listener_manager.h]: And  no new changes are allowed to listeners after the `stopListeners` call - that would be behavioural change if we merge, but should be OK I guess?

**Comment by ramaraochavali** [comment]
@mattklein123 assuming that there is no need to maintain previous API, I just merged both the APIs and also introduced `StopListenerSelector` with a `ListenerDirection` enum - This would allow us to expand to other criteria in future if needed. PTAL.

File: include/envoy/server/listener_manager.h [hunk 14]
```diff
@@ -102,6 +102,20 @@ class ListenerComponentFactory {
  */
 class ListenerManager {
 public:
+  // Criteria for selecting listeners to stop.
+  struct StopListenerSelector {
+
+    enum class ListenerDirection {
```
Comment (mattklein123) [inline include/envoy/server/listener_manager.h]: nit: I don't think you need the outer struct. Can you just call this enum `StopListenersType` or something and use that as the param to the `stopListeners` function?

Comment (ramaraochavali) [reply include/envoy/server/listener_manager.h]: As I mentioned the outer struct was added to support for other criteria in future like name - I am OK removing that for now and adding it later when it is needed, if you think that is best

Comment (mattklein123) [reply include/envoy/server/listener_manager.h]: Yeah let's remove for now.

**Comment by ramaraochavali** [comment]
@mattklein123  removed the outer struct. PTAL

File: docs/root/operations/admin.rst [hunk 15]
```diff
@@ -203,6 +203,22 @@ modify different aspects of the server:
...
+   :ref:`Drains <arch_overview_draining>` all listeners.
+
+   .. http:post:: /drain_listeners?inboundonly
+
+   :ref:`Drains <arch_overview_draining>` all inbound listeners. `traffic_direction` field in :ref:`Listener <envoy_api_msg_Listener>` 
+   is used to determine whether a listener is inbound or outbound.
+
+.. attention::
+
+   This operation directly stops the matched listeners on workers. Once listeners in a given traffic direction are stopped, listener additions
```
Comment (mattklein123) [inline docs/root/operations/admin.rst]: nit: reflow at 100 columns here and above.

File: include/envoy/server/listener_manager.h [hunk 16]
```diff
@@ -102,6 +102,16 @@ class ListenerComponentFactory {
  */
 class ListenerManager {
 public:
+  // Indicates listeners to stop.
+  enum class StopListenersType {
+    // Listeners in the inbound direction are only stopped.
+    InboundOnly,
+    // All listeners are stopped.
+    All,
+    // Listeners are not stopped by direction.
```
Comment (mattklein123) [inline include/envoy/server/listener_manager.h]: It's a little odd to me that we have a stop type of "none" when passed as an API parameter. Can we just remove this and use an optional for the internal state if nothing is stopped yet?

File: source/server/http/admin.cc (hunk 11 above)
Comment (mattklein123) [reply source/server/http/admin.cc]: +1 I'm not sure if it's worth it. Can we remove this and then make the function return void again?

**Comment by ramaraochavali** [comment]
@mattklein123 addressed the feedback. PTAL

File: include/envoy/server/listener_manager.h (hunk 16 above)
Comment (ramaraochavali) [reply include/envoy/server/listener_manager.h]: I have added it earlier to support other criteria and that time direction can be None. But now it is not needed. removed

File: source/server/listener_manager_impl.cc [hunk 17]
```diff
@@ -553,9 +561,24 @@ void ListenerManagerImpl::startWorkers(GuardDog& guard_dog) {
...
 }
 
-void ListenerManagerImpl::stopListeners() {
-  for (const auto& worker : workers_) {
-    worker->stopListeners();
+void ListenerManagerImpl::stopListeners(StopListenersType listeners_type) {
+  listeners_type_ = listeners_type;
+  for (Network::ListenerConfig& listener : listeners()) {
+    for (const auto& worker : workers_) {
+      if (listeners_type_.has_value() &&
```
Comment (mattklein123) [inline source/server/listener_manager_impl.cc]: This must have a value, right? You set it above.

File: source/server/listener_manager_impl.h [hunk 18]
```diff
@@ -191,6 +199,7 @@ class ListenerManagerImpl : public ListenerManager, Logger::Loggable<Logger::Id:
   std::list<DrainingListener> draining_listeners_;
   std::list<WorkerPtr> workers_;
   bool workers_started_{};
+  absl::optional<StopListenersType> listeners_type_;
```
Comment (mattklein123) [inline source/server/listener_manager_impl.h]: nit: stop_listeners_type_

File: test/integration/integration_admin_test.cc [hunk 19]
```diff
@@ -448,6 +453,41 @@ TEST_P(IntegrationAdminTest, Admin) {
...
+  // Validate that the drain_listeners is processed correctly.
+  response = IntegrationUtil::makeSingleRequest(lookupPort("admin"), "POST", "/drain_listeners", "",
+                                                downstreamProtocol(), version_);
+  EXPECT_TRUE(response->complete());
+  EXPECT_EQ("200", response->headers().Status()->value().getStringView());
+  EXPECT_EQ("text/plain; charset=UTF-8", ContentType(response));
+  EXPECT_EQ("OK\n", response->body());
+}
+
+// Validates that the inboundonly drains inbound listeners.
```
Comment (mattklein123) [inline test/integration/integration_admin_test.cc]: How is this actually validated? Is there a stat you can wait on, etc? I'm not sure how this test is different in practice from the one above unless you verify that the drain actually happens via some mechanism?

File: test/integration/integration_test.cc [hunk 20]
```diff
@@ -92,6 +92,39 @@ TEST_P(IntegrationTest, PerWorkerStatsAndBalancing) {
...
+                                          {":path", "/test/long/url"},
+                                          {":scheme", "http"},
+                                          {":authority", "host"}};
+  IntegrationStreamDecoderPtr response = codec_client_->makeHeaderOnlyRequest(request_headers);
+  waitForNextUpstreamRequest(0);
+  fake_upstreams_[0]->set_allow_unexpected_disconnects(true);
+
+  upstream_request_->encodeHeaders(default_response_headers_, false);
+
+  // Invoke drain listeners endpoint and validate that the listener is actually draining.
```
Comment (mattklein123) [inline test/integration/integration_test.cc]: How have you verified it is draining?

File: test/integration/integration_admin_test.cc (hunk 19 above)
Comment (ramaraochavali) [reply test/integration/integration_admin_test.cc]: There is no stat to validate- it just validates that call is done with out errors when we pass the additional argument. I will correct the comment though

Comment (mattklein123) [reply test/integration/integration_admin_test.cc]: Why can't you validate that draining/deletion actually happens? I'm not sure the point of this test otherwise? How is this any different from the above tests?

File: test/integration/integration_test.cc (hunk 20 above)
Comment (ramaraochavali) [reply test/integration/integration_test.cc]: it validates that after the listener drain is invoked, we can send some data. The drain part is validated in the listener_manager_impl_test by expecting on stopListener call on worker

File: source/server/listener_manager_impl.cc (hunk 17 above)
Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: Yeah :-). Morning needs more coffee..

File: test/integration/integration_admin_test.cc (hunk 19 above)
Comment (ramaraochavali) [reply test/integration/integration_admin_test.cc]: Oh. Sorry, This test is not required. I will remove this.

Comment (ramaraochavali) [reply test/integration/integration_admin_test.cc]: Also enhanced it to cover few more cases. We should be good.

File: source/server/listener_manager_impl.cc [hunk 21]
```diff
@@ -553,9 +561,23 @@ void ListenerManagerImpl::startWorkers(GuardDog& guard_dog) {
   }
 }
 
-void ListenerManagerImpl::stopListeners() {
-  for (const auto& worker : workers_) {
-    worker->stopListeners();
+void ListenerManagerImpl::stopListeners(StopListenersType listeners_type) {
```
Comment (mattklein123) [inline source/server/listener_manager_impl.cc]: nit: s/listeners_type/stop_listeners_type (this will fall clang-tidy at some point)

File: test/integration/integration_admin_test.cc [hunk 22]
```diff
@@ -123,6 +123,11 @@ std::string ContentType(const BufferingStreamDecoderPtr& response) {
 } // namespace
 
 TEST_P(IntegrationAdminTest, Admin) {
+  config_helper_.addConfigModifier([&](envoy::config::bootstrap::v2::Bootstrap& bootstrap) -> void {
```
Comment (mattklein123) [inline test/integration/integration_admin_test.cc]: Does this add anything to this test? We aren't verifying any behavior below AFAICT? This is why I have been asking if there is any way that we can actually verify that we are draining a listener. Even if there is no stat could we potentially look for an info log either in this test or in the proper integration test?

File: test/integration/integration_test.cc [hunk 23]
```diff
@@ -92,6 +92,39 @@ TEST_P(IntegrationTest, PerWorkerStatsAndBalancing) {
   check_listener_stats(0, 1);
 }
 
+// Validates that the drain actually drains the listeners.
```
Comment (mattklein123) [inline test/integration/integration_test.cc]: I'm sorry to be pedantic, but I still don't see how this test actually verifies that we have drained the listeners.

If the drain timeout is say 1s, could we somehow verify the listener is actually stopped at the end of the test? Either via a stat? Trying to connect and it failing? Getting a connection-close header on the response? Something?

File: source/server/listener_manager_impl.cc (hunk 21 above)
Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: sorry. missed this. Will change

File: test/integration/integration_admin_test.cc (hunk 22 above)
Comment (ramaraochavali) [reply test/integration/integration_admin_test.cc]: Yes it does. It modifies the existing listener and sets the traffic direction as OUTBOUND. I have added the stat now - Now it should be clear. PTAL

Comment (ramaraochavali) [reply test/integration/integration_admin_test.cc]: I also brought back the second test, because with stat we can validate that behaviour as well.

File: test/integration/integration_test.cc (hunk 23 above)
Comment (ramaraochavali) [reply test/integration/integration_test.cc]: I have added a new stat and verified here. PTAL. Just to be clear on drain timeout, drain timeout does not have any impact here - listener would be stopped immediately on workers when `drain_listeners` endpoint is called.

File: docs/root/configuration/listeners/stats.rst [hunk 24]
```diff
@@ -70,3 +70,4 @@ statistics. Any ``:`` character in the stats name is replaced with ``_``.
    total_listeners_warming, Gauge, Number of currently warming listeners
    total_listeners_active, Gauge, Number of currently active listeners
    total_listeners_draining, Gauge, Number of currently draining listeners
+   total_listeners_stopped, Gauge, Number of listeners that were stopped
```
Comment (mattklein123) [inline docs/root/configuration/listeners/stats.rst]: This seems like this should just be a counter? If not don't you need to keep track of stopped listeners that have been removed, etc.?

File: test/server/listener_manager_impl_test.cc [hunk 25]
```diff
@@ -1111,6 +1111,203 @@ name: foo
   checkStats(2, 1, 2, 0, 0, 0);
 }
 
+// Validates that StopListener functionality works correctly when only inbound listeners are
```
Comment (mattklein123) [inline test/server/listener_manager_impl_test.cc]: Please test the stopped stat (counter?) here also.

**Comment by ramaraochavali** [comment]
@mattklein123 moved to counter and added verification in listener_manager_impl_test as well. PTAL
